### PR TITLE
keep cursor stable & fix go to symbol as terminal accessibility buffer dynamically updates / resizes

### LINKED ISSUE #177923: Cannot navigate to commands as buffer dynamically updates
From @Tyriar:

- Fill the scrollback
- Run some command that outputs over time
- Focus accessible buffer
- Navigate commands via the quick pick, _bufferToEditorIndex will point at the original stale values

### LINKED ISSUE #178695: dynamic content can be added to the terminal accessible buffer twice
Screenshot 2023-03-30 at 8 20 34 AM

### LINKED ISSUE #178703: cursor position is not preserved when data is added dynamically to the accessible buffer
run a command like:

```
#!/bin/bash

count=0

while true; do
  echo $count
  count=$((count+1))
  sleep 1
done
```

`shift+tab` to focus the accessible buffer and click somewhere in the there so the cursor position is stable
🐛 it moves when a new line is added

### LINKED ISSUE #179186: scroll behavior in terminal accessible buffer as it dynamically updates is choppy
Repro:

Enter the accessible buffer with some dynamically outputting command and scrollback. 

1. Use the mouse wheel to scroll and see that the editor scrolls to the bottom as new content is added
2. Using arrow keys, reposition the cursor. It feels jumpy

### LINKED ISSUE #179273: incorrect accessible buffer contribution is accessed when there are several
https://user-images.githubusercontent.com/29464607/230130129-b7669a63-a7d4-46e1-b632-04697a158b12.mov

### PR DESCRIPTION
**Pull request by meganrogge**
fixes #178703 
fixes [#178695](https://github.com/microsoft/vscode/issues/178695)
fixes #177923
fixes #179186 
fixes #179273

### CODE REVIEW

File: src/vs/workbench/contrib/terminalContrib/accessibility/browser/bufferContentTracker.ts [hunk 1]
```diff
@@ -30,28 +30,63 @@ export class BufferContentTracker {
...
+			}
+		}
 	}
 
 	private _removeViewportContent(): void {
 		if (this._lines.length && this._lastCachedMarker) {
 			// remove previous viewport content in case it has changed
 			let i = 0;
-			while (i <= this._priorViewportLineCount) {
+			while (i < (this._priorViewportLineCount < this._xterm.raw.rows ? this._priorViewportLineCount - 1 : this._priorViewportLineCount)) {
```
Comment (meganrogge) [inline src/vs/workbench/contrib/terminalContrib/accessibility/browser/bufferContentTracker.ts:76]: this was the crux of the issue and doesn't make much sense to me

**Comment by meganrogge** [comment]
just noticed an issue where go to symbol doesn't contain the commands that are in the scroll back 😢

File: src/vs/workbench/contrib/terminalContrib/accessibility/browser/terminalAccessibleBuffer.ts [hunk 2]
```diff
@@ -130,6 +132,13 @@ export class AccessibleBufferWidget extends DisposableStore {
 					break;
 			}
 		}));
+		this.add(this._editorWidget.onDidChangeCursorPosition((c) => {
+			if (c.source === 'mouse' || c.source === 'keyboard') {
+				this._cursorPosition = { lineNumber: c.position.lineNumber, column: c.position.column };
+			} else if (this._cursorPosition) {
+				this._resetPosition();
+			}
+		}));
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminalContrib/accessibility/browser/terminalAccessibleBuffer.ts:141]: @rebornix are we doing something wrong here by applying edits to the model as changes come in? We're needing to jump through a bunch of hoops to keep the cursor and scroll position when changes happen in the terminal buffer and we add them to the monaco instance.

Comment (rebornix) [reply src/vs/workbench/contrib/terminalContrib/accessibility/browser/terminalAccessibleBuffer.ts:141]: I think we might run into trouble in the future by listening to cursor position change and then resetting it. From what @Tyriar shared about how we ran into this problem, there are two alternatives:

* Before setting new value to the model, save the view state (which includes the cursor state). Once model is updated, restore the view state.
* Or we do a diff of the content, and apply edits incrementally, like what we did in text file editor https://insiders.vscode.dev/github/microsoft/vscode/blob/rebornix/integral-monkey/src/vs/editor/common/services/modelService.ts#L396-L397

Comment (meganrogge) [reply src/vs/workbench/contrib/terminalContrib/accessibility/browser/terminalAccessibleBuffer.ts:141]: I have tried many approaches and believe this is the best solution. Let's see if there are issues before rethinking the strategy.

**Review by Tyriar — APPROVED** [review]

**Pull request merged**